File: analyser/traverser.go
package analyser

import (
	"github.com/john-nguyen09/go-phpparser/phrase"
)

func Traverse(node phrase.AstNode, analyser *Analyser) {
	if p, ok := node.(*phrase.Phrase); ok {
		analyser.Preorder(p)

		for _, child := range p.Children {
			Traverse(child, analyser)
		}

		analyser.Postorder(p)
	}
}

File: src/traverser.ts
import "reflect-metadata";
import { injectable } from "inversify";
import { TreeNode, isPhrase } from "./util/parseTree";
import { Phrase } from "php7parser";

@injectable()
export class Traverser {
    private visitors: Visitor[] = [];

    traverse(rootNode: TreeNode, visitors: Visitor[]): void {
        this.visitors = visitors;
        this.realTraverse(rootNode, []);
        this.visitors = [];
    }

    private realTraverse(node: TreeNode, spine: Phrase[]) {
        for (let visitor of this.visitors) {
            visitor.preorder(node, spine);
        }

        if (isPhrase(node)) {
            spine.push(node);
            for (let child of node.children) {
                this.realTraverse(child, spine);
            }
            spine.pop();
        }

        for (let visitor of this.visitors) {
            if (visitor.postorder != undefined) {
                visitor.postorder(node);
            }
        }
    }
}

export interface Visitor {
    preorder(node: TreeNode, spine: TreeNode[]): void;
    postorder?(node: TreeNode): void;
}